src/modules/sensors/vehicle_air_data/VehicleAirData.cpp

```diff
@@ -0,0 +1,280 @@
+			_sensor_sub[_selected_sensor_sub_index].registerCallback();
+		}
+	}
+
+	if ((_selected_sensor_sub_index >= 0) && updated[_selected_sensor_sub_index]) {
+		ParametersUpdate();
+		SensorCorrectionsUpdate();
+
+		const sensor_baro_s &baro = _last_data[_selected_sensor_sub_index];
+
+		// populate vehicle_air_data with primary baro and publish
+		vehicle_air_data_s out{};
+		out.timestamp_sample = baro.timestamp; // TODO: baro.timestamp_sample;
+		out.baro_device_id = baro.device_id;
+		out.baro_temp_celcius = baro.temperature;
+
+		// Convert from millibar to Pa and apply temperature compensation
+		out.baro_pressure_pa = (100.0f * baro.pressure - _offset[_selected_sensor_sub_index]) *
+				       _scale[_selected_sensor_sub_index];
+
+		// calculate altitude using the hypsometric equation
+		static constexpr float T1 = 15.0f - CONSTANTS_ABSOLUTE_NULL_CELSIUS; // temperature at base height in Kelvin
+		static constexpr float a = -6.5f / 1000.0f; // temperature gradient in degrees per metre
+
+		// current pressure at MSL in kPa (QNH in hPa)
+		const float p1 = _param_sens_baro_qnh.get() * 0.1f;
+
+		// measured pressure in kPa
+		const float p = out.baro_pressure_pa * 0.001f;
+
+		/*
+		 * Solve:
+		 *
+		 *     /        -(aR / g)     \
+		 *    | (p / p1)          . T1 | - T1
+		 *     \                      /
+		 * h = -------------------------------  + h1
+		 *                   a
+		 */
+		out.baro_alt_meter = (((powf((p / p1), (-(a * CONSTANTS_AIR_GAS_CONST) / CONSTANTS_ONE_G))) * T1) - T1) / a;
+
+		// calculate air density
+		// estimate air density assuming typical 20degC ambient temperature
+		// TODO: use air temperature if available (differential pressure sensors)
+		static constexpr float pressure_to_density = 1.0f / (CONSTANTS_AIR_GAS_CONST * (20.0f -
+				CONSTANTS_ABSOLUTE_NULL_CELSIUS));
+
+		out.rho = pressure_to_density * out.baro_pressure_pa;
+
+		out.timestamp = hrt_absolute_time();
+		_vehicle_air_data_pub.publish(out);
+	}
+
+	// check failover and report
+	if (_last_failover_count != _voter.failover_count()) {
+		uint32_t flags = _voter.failover_state();
+		int failover_index = _voter.failover_index();
+
+		if (flags == DataValidator::ERROR_FLAG_NO_ERROR) {
+			if (failover_index != -1) {
+				// we switched due to a non-critical reason. No need to panic.
+				PX4_INFO("sensor_baro switch from #%i", failover_index);
+			}
+
+		} else {
+			if (failover_index != -1) {
+				mavlink_log_emergency(&_mavlink_log_pub, "sensor_baro:#%i failed: %s%s%s%s%s!, reconfiguring priorities",
+						      failover_index,
+						      ((flags & DataValidator::ERROR_FLAG_NO_DATA) ? " OFF" : ""),
+						      ((flags & DataValidator::ERROR_FLAG_STALE_DATA) ? " STALE" : ""),
+						      ((flags & DataValidator::ERROR_FLAG_TIMEOUT) ? " TIMEOUT" : ""),
+						      ((flags & DataValidator::ERROR_FLAG_HIGH_ERRCOUNT) ? " ERR CNT" : ""),
+						      ((flags & DataValidator::ERROR_FLAG_HIGH_ERRDENSITY) ? " ERR DNST" : ""));
```

Comment:
Can we spell out `CNT` and `DNST` otherwise a pilot does not understand it.

Reply:
Once I finish splitting `sensors` into separate WorkItems per data type (next is mag, then IMU) let's take a pass and cleanup this output. I'm not sure that we should be sending any of this via mavlink text when we have SYS_STATUS flags.